### DISCUSSION TITLE
Does the Reactor also transmit responses in addition to buffering requests?

### DISCUSSION THREAD

**Discussion by bbascarevic** [puma/puma#3090] [Q&A] [answered]
If I understand the docs/architecture.md document correctly, Puma protects from slow clients by employing an async nio reactor that reads requests fully before handing them off to available worker threads for processing. Slow clients won't affect the worker thread pool and the number of concurrent slow clients is probably only limited by the available memory. But once the request is processed and the response needs to be transmitted, is this done by the worker thread or is there a nio reactor that handles transmitting responses to clients at their pace?

Also an additional question, is the entire request, potential large body included, read by the reactor before the handoff with no exceptions?

**Comment by nateberkopec** [comment, marked as answer]
Responses are not buffered in Puma, so the ThreadPool worker thread will send them to the client directly.

The entire request is read by the reactor unless `queue_requests` is `false`.

**Reply by bbascarevic** [reply]
@nateberkopec Thank you very much for the answer. Just out of curiosity, is this something that's simply not implemented but could/should be, or is it a non-issue in practice? Could the host OS/network stack be configured in some way to help with this if it's potentially an issue?

**Reply by stevenharman** [reply]
Does this mean that Puma is still susceptible to the "slow client problem" since ThreadPool workers can be hung up sending the request to the client?

**Reply by nateberkopec** [reply]
I asked Evan: the answer is that it's simply not implemented.

> Does this mean that Puma is still susceptible to the "slow client problem"

I think that would have to proven with a proof of concept first. Sending is quite different from receiving, as we are not actually waiting in the same way for the client to _do something_.

**Reply by stevenharman** [reply]
> Sending is quite different from receiving, as we are not actually waiting in the same way for the client to _do something_.

Maybe? I'm not sure I follow exactly what you mean. We _are waiting_ for the client to drain the TCP receive buffer, yeah? So a client reading the buffer slowly means the Puma worker thread is tied up with just that client the whole time, right? (This assumes there's no reverse-proxy or something in front, doing the response buffering. e.g., Heroku's Common Runtime routers will buffer up to 1MiB of response per connection.)

**Reply by nateberkopec** [reply]
You're right, my understanding here was incomplete. 

This looks like a good resource (and tool) for understanding slow read attacks https://github.com/shekyan/slowhttptest/wiki There are some ideas for mitigation here which do not actually require an event loop either.

**Reply by nateberkopec** [reply]
So then our potential user stories are:

1. Mitigate slow read attacks by putting responses in the Reactor loop.
2. Mitigate slow read attacks through other means, e.g. limits on connection lifetime, limits on acceptable window sizes

**Reply by Vuta** [reply]
Normally people would use Puma behind a reverse proxy like Nginx or Apache, and those proxies have their own mechanism to mitigate this problem too. So, not sure if it's worth implementing these in Puma itself?

**Reply by nateberkopec** [reply]
We've always supported the "without a reverse proxy" usecase. It's one of the things that makes this project different than, say, Unicorn, which does explicitly say in their Readme that you should use a reverse proxy. It's why we have SSL and request buffering. So, I do think it's something we can support.

**Reply by stevenharman** [reply]
> We've always supported the "without a reverse proxy" usecase.

Yep! And I've always appreciated that. In fact, we're looking to move from an nginx + Unicorn setup to Puma in our Heroku-based deployment. Part of that migration strategy includes mitigating potential thread-safety issues in our code. To do that, we're going to use a Dyno-by-Dyno feature toggle that will boot using Puma in a multi-process, single-threaded mode. Which led me to looking into how Puma handles slow-reading clients, and this discussion. 😄

@nateberkopec Those options make sense to me. My gut-reaction would be option one as it limits the configuration needed to protect against slow-reading clients. But I'm not familiar enough with the technical/implementation tradeoffs of that option to have an opinion.

**Reply by nateberkopec** [reply]
This is now issue #3110

**Reply by stevenharman** [reply]
Thank you, @nateberkopec!
